File: hackerrank/algorithms/graph-theory/primsmstsub/primsmstsub.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# "Prim's (MST) : Special Subtree" on HackerRank
# https://www.hackerrank.com/challenges/primsmstsub
# Using a binary minheap + map data structure.

$VERBOSE = 1

# A heap + map data structure for Prim's and Dijkstra's algorithms.
class PrimHeap
  def initialize
    @heap = [] # index -> (key, value)
    @map = {} # key -> index
  end

  def empty?
    @heap.empty?
  end

  def size
    @heap.size
  end

  def push_or_decrease(key, value)
    child = @map[key]

    if child.nil?
      sift_up(size, make_entry(key, value))
    elsif value < @heap[child].value
      sift_up(child, make_entry(@heap[child].key, value))
    end
  end

  def pop
    raise "can't pop from empty heap" if empty?

    key, value = *@heap.first
    @map.delete(key)

    if @map.empty?
      @heap.clear
    else
      parent_entry = @heap.pop
      sift_down(0, parent_entry)
    end

    [key, value]
  end

  private

  def sift_up(child, child_entry)
    while child.positive?
      parent = (child - 1) / 2
      break if @heap[parent].value <= child_entry.value

      set(child, @heap[parent])
      child = parent
    end

    set(child, child_entry)
  end

  def sift_down(parent, parent_entry)
    loop do
      child = pick_child(parent)
      break if child.nil? || parent_entry.value <= @heap[child].value

      set(parent, @heap[child])
      parent = child
    end

    set(parent, parent_entry)
  end

  def pick_child(parent)
    left = parent * 2 + 1
    return nil if left >= size

    right = left + 1
    right == size || @heap[left].value <= @heap[right].value ? left : right
  end

  def set(index, entry)
    @heap[index] = entry
    @map[entry.key] = index
    nil
  end

  def make_entry(key, value)
    Entry.new(key, value).freeze
  end

  Entry = Struct.new(:key, :value)
  private_constant :Entry
end

# A weighted undirected graph.
class Graph
  def initialize(vertex_count)
    @adj = Array.new(vertex_count) { [] }
  end

  def vertex_count
    @adj.size
  end

  def add_edge(u, v, weight)
    raise 'endpoint vertex out of range' unless exists?(u) && exists?(v)

    @adj[u] << make_out_edge(v, weight)
    @adj[v] << make_out_edge(u, weight)
    nil
  end

  # Sums the weights of edges in the component that contains the start vertex,
  # using Prim's algorithm.
  def prim_mst(start)
    raise 'start vertex out of range' unless exists?(start)

    total_cost = 0
    vis = [false] * vertex_count
    heap = PrimHeap.new
    heap.push_or_decrease(start, 0)

    until heap.empty?
      src, src_cost = heap.pop
      vis[src] = true
      total_cost += src_cost

      @adj[src].each do |out_edge|
        dest, weight = *out_edge
        heap.push_or_decrease(dest, weight) unless vis[dest]
      end
    end

    total_cost
  end

  private

  def exists?(vertex)
    vertex.between?(0, vertex_count - 1)
  end

  def make_out_edge(dest, weight)
    OutEdge.new(dest, weight).freeze
  end

  OutEdge = Struct.new(:dest, :weight)
  private_constant :OutEdge
end

def read_record
  gets.split.map(&:to_i)
end

def build_graph
  vertex_count, edge_count = read_record
  graph = Graph.new(vertex_count + 1) # +1 for 1-based indexing
  edge_count.times { graph.add_edge(*read_record) }
  graph
end

if $PROGRAM_NAME == __FILE__
  graph = build_graph
  start = gets.to_i
  puts graph.prim_mst(start)
end

File: hackerrank/algorithms/graph-theory/primsmstsub/primsmstsub.cpp
// Prim's (MST): Special Subtree
// https://www.hackerrank.com/challenges/primsmstsub
// Using std::priority_queue.

#include <iostream>
#include <queue>
#include <vector>

namespace {
    // Visitation state for vertex (to avoid std::vector<bool>).
    enum Vis : bool { not_done, done };

    // Information on an outward neighbor in a weighted undirected graph.
    struct OutEdge {
        int dest;
        int weight;
    };

    // Priority-queue entry representing best cost so far to reach a vertex.
    struct Entry {
        int vertex;
        int cost;
    };

    // Create an initially empty priority queue for use in Prim's algorithm.
    auto make_priority_queue() noexcept
    {
        constexpr auto compare = [](const Entry& lhs,
                                    const Entry& rhs) noexcept
        {
            return lhs.cost > rhs.cost;
        };

        return std::priority_queue<
            Entry,
            std::vector<Entry>,
            decltype(compare)
        >{compare};
    }

    // A weighted undirected graph.
    class Graph {
    public:
        // Make a graph with vertices 0, ..., vertex_count - 1, no edges yet.
        Graph(int vertex_count) noexcept;

        // Add an edge to the graph.
        void add_edge(int u, int v, int weight) noexcept;

        // Compute the weight of a minimum spanning tree with Prim's algorithm.
        // The MST is found from the given start vertex.
        [[nodiscard]] int mst_weight(int start) const noexcept;

    private:
        std::vector<std::vector<OutEdge>> adj_;
    };

    Graph::Graph(const int vertex_count) noexcept : adj_(vertex_count)
    {
    }

    inline void
    Graph::add_edge(const int u, const int v, const int weight) noexcept
    {
        adj_[u].push_back({v, weight});
        adj_[v].push_back({u, weight});
    }

    int Graph::mst_weight(const int start) const noexcept
    {
        constexpr auto unseen = -1;

        auto pq = make_priority_queue();
        pq.push({start, 0});
        auto costs = std::vector<int>(size(adj_), unseen);
        costs[start] = 0;
        auto vis = std::vector<Vis>(size(adj_), Vis::not_done);
        auto total_weight = 0;

        while (!empty(pq)) {
            const auto [src, src_cost] = pq.top();
            pq.pop();
            if (vis[src] == Vis::done) continue;

            vis[src] = Vis::done;
            total_weight += src_cost;

            for (const auto [dest, weight] : adj_[src]) {
                if (vis[dest] == Vis::done) continue;

                if (costs[dest] == unseen || costs[dest] > weight) {
                    costs[dest] = weight;
                    pq.push({dest, weight});
                }
            }
        }

        return total_weight;
    }

    [[nodiscard]] Graph read_graph() noexcept
    {
        auto vertex_count = 0, edge_count = 0;
        std::cin >> vertex_count >> edge_count;

        auto graph = Graph{vertex_count + 1}; // +1 for 1-based indexing.

        while (edge_count-- > 0) {
            auto u = 0, v = 0, weight = 0;
            std::cin >> u >> v >> weight;
            graph.add_edge(u, v, weight);
        }

        return graph;
    }
}

int main()
{
    std::ios_base::sync_with_stdio(false);

    const auto graph = read_graph();
    auto start = 0;
    std::cin >> start;
    std::cout << graph.mst_weight(start) << '\n';
}
